subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -500,6 +530,185 @@ static bool dtls_is_peer_addr_valid(struct tls_context *context,
+		entry->session = NULL;
+	}
+
+	(void)mbedtls_ssl_session_save(session, NULL, 0, &session_len);
+
+	entry->session = mbedtls_calloc(1, session_len);
+	if (entry->session == NULL) {
+		NET_ERR("Failed to allocate session buffer.");
+		return -ENOMEM;
+	}
+
+	ret = mbedtls_ssl_session_save(session, entry->session, session_len,
+				       &session_len);
+	if (ret < 0) {
+		NET_ERR("Failed to serialize session, err: 0x%x.", -ret);
+		mbedtls_free(entry->session);
+		entry->session = NULL;
+		return -ENOMEM;
+	}
+
+	entry->session_len = session_len;
+	entry->timestamp = k_uptime_get();
+	memcpy(&entry->peer_addr, peer_addr, sizeof(*peer_addr));
+
+	return 0;
+}
+
+static int tls_session_get(const struct sockaddr *peer_addr,
+			   mbedtls_ssl_session *session)
+{
+	struct tls_session_cache *entry = NULL;
+	int ret;
+
+	for (int i = 0; i < ARRAY_SIZE(client_cache); i++) {
+		if (client_cache[i].session != NULL &&
+		    peer_addr_cmp(&client_cache[i].peer_addr, peer_addr)) {
+			entry = &client_cache[i];
+			break;
+		}
+	}
+
+	if (entry == NULL) {
+		return -ENOENT;
+	}
+
+	ret = mbedtls_ssl_session_load(session, entry->session,
+				       entry->session_len);
+	if (ret < 0) {
+		/* Discard corrupted session data. */
+		mbedtls_free(entry->session);
+		entry->session = NULL;
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void tls_session_store(struct tls_context *context,
+			      const struct sockaddr *addr,
+			      socklen_t addrlen)
+{
+	mbedtls_ssl_session session;
+	struct sockaddr peer_addr = { 0 };
+	int ret;
+
+	if (!context->options.cache_enabled) {
+		return;
+	}
+
+	memcpy(&peer_addr, addr, addrlen);
+	mbedtls_ssl_session_init(&session);
+
+	ret = mbedtls_ssl_get_session(&context->ssl, &session);
+	if (ret < 0) {
+		NET_ERR("Failed to obtain session for %p", context);
+		goto exit;
+	}
+
+	ret = tls_session_save(&peer_addr, &session);
+	if (ret < 0) {
+		NET_ERR("Failed to save session for %p", context);
+	}
+
+exit:
+	mbedtls_ssl_session_free(&session);
+}
+
+static void tls_session_restore(struct tls_context *context,
+				const struct sockaddr *addr,
+				socklen_t addrlen)
+{
+	mbedtls_ssl_session session;
+	struct sockaddr peer_addr = { 0 };
+	int ret;
+
+	if (!context->options.cache_enabled) {
+		return;
+	}
+
+	memcpy(&peer_addr, addr, addrlen);
+	mbedtls_ssl_session_init(&session);
+
+	ret = tls_session_get(&peer_addr, &session);
+	if (ret < 0) {
+		NET_DBG("Session not found for %p", context);
+		goto exit;
+	}
+
+	ret = mbedtls_ssl_set_session(&context->ssl, &session);
+	if (ret < 0) {
+		NET_ERR("Failed to set session for %p", context);
+	}
+
+exit:
+	mbedtls_ssl_session_free(&session);
+}
+
+static void tls_session_purge(void)
+{
+	tls_session_cache_reset();
+
+#if defined(MBEDTLS_SSL_CACHE_C)
+	mbedtls_ssl_cache_free(&server_cache);
+	mbedtls_ssl_cache_init(&server_cache);
+#endif
+}
+
+static inline int time_left(uint32_t start, uint32_t timeout)
+{
+	uint32_t elapsed = k_uptime_get_32() - start;
```

Comment:
Do we need to worry about overflow here? This will wrap after about 49 days.

Reply:
I don't think this would ever be the case - the function is used with a timeout value provided by mbed TLS, which is 32-bit.